### PR TITLE
LL-277

### PR DESCRIPTION
**Pull request by KevinLemon112**
# Description

In this new PR/Ticket, unit tests for the Publish and Subscribe features for the ActiveMQ notification service have been created and tested. 

Subsequent # (issue): LL-277

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Steps (For Local Host):

1. Be in the server/src/notification/publish OR server/src/notification/subscribe directories to test each unit test separately
2. Run "npm i" if needed in each directory (as well as the root directory)
4. Run "npx jest" in the terminal to test the unit test for the directory you are in.
 
# Checklist (when relevant):

- [x] My code follows the style guidelines of this project

[TypeScript](https://google.github.io/styleguide/tsguide.html)
[Python](https://peps.python.org/pep-0008/)
[Schema](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/forefront-2010/ee652242(v=vs.100))
- [x] I have included a reviewer on the pull request [*Replace this text with the reviewer*]
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/notification/publish/publish.test.ts [hunk 1]
```diff
@@ -1,28 +1,121 @@
...
+
+      client.on('error', (error: Error) => {
+        log(error);
+        res.status(500).json({ error: 'Internal Server Error' });
+      });
+    });
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
```
Comment (pogi7) [inline src/notification/publish/publish.test.ts:92]: @KevinLemon112 I believe you need to close the express server when the tests finish.  Look at this stackoverflow for more info.
https://stackoverflow.com/questions/59011575/how-to-close-express-server-inside-jest-afterall-hook

File: src/notification/subscribe/subscribe.test.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    expect(response.status).toBe(200);
+    expect(response.body).toEqual({ message: 'Subscriber is running' });
+  });
+
+  it('should log messages received from the broker', () => {
+    const logSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
+
+    // Ensure the MQTT client is set up and used
+    const connectMock = mqtt.connect as jest.Mock;
```
Comment (pogi7) [inline src/notification/subscribe/subscribe.test.ts:83]: @KevinLemon112 The connection looks like but I don't think any notifications are being mocked.  Look at this stackoverflow on how to mock notifications for mqtt.
https://stackoverflow.com/questions/43290317/jest-mocking-simple-mock

File: src/notification/publish/publish.test.ts [hunk 3]
```diff
@@ -1,28 +1,141 @@
...
+        done(err);
+      } else {
+        console.log("Server closed successfully.");
+        done();
+      }
+    });
+  });
+
+  it('should publish a message and return the event object', async () => {
+    const response = await request(app).get('/publish/12345');
```
Comment (pogi7) [inline src/notification/publish/publish.test.ts:108]: @KevinLemon112 I think this request should be mocked.  Please refer to this stack overflow article as an example of how to mock requests to an ExpressJS Server.
https://stackoverflow.com/a/72674717

**Comment by Verus20** [comment]
@KevinLemon112 the tests work nice job!  **recommendation:** In the `How has this been tested` stage, I'd recommending adding a step to run npm i from the root directory of the server (i.e the folder before src)
image

This is because I was running into errors with jest and then running npm i from the root folder got the tests to work

File: src/notification/subscribe/subscribe.test.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    //   if (messageHandler) {
+    //     messageHandler('test-topic', Buffer.from('test-message')); // Trigger the message event
+    //   }
+    // }
+
+    const expectedMessage = 'Received message from topic test-topic: test-message';
+
+    // FIXME: This code returns [["connectMock call count after setup:", 0], ["Actual calls to log:", [Circular]]]
+    // const actualCalls = logSpy.mock.calls.map((call: any[]) => call.join(' '));
+    // logSpy.mockRestore();
```
Comment (Verus20) [inline src/notification/subscribe/subscribe.test.ts:97]: **question:** is this commented out code still needed?  If not, could it be removed?

Comment (Verus20) [reply src/notification/subscribe/subscribe.test.ts:97]: resolved: this can be used at later time

**Review by Verus20 — APPROVED** [review]

**Pull request merged**